### ISSUE TITLE
Run clean before compiling target during `elixir_make.precompile`

### ISSUE THREAD

**Issue by akash-akya** [elixir-lang/elixir_make#67]
In my case, to compile the NIF I need few supporting libraries. These libraries are configured as Makefile target prerequisite, so that when I try to build NIF, it automatically fetches these files and builds the NIF. These libraries are platform specific, so I need to fetch it different files based for different arch/os/abi.

### Issue
`mix elixir_make.precompile` tries to build the package for all supported platforms. But during that, it does not `clean` before starting the next target compilation. Because of that, I am not able to create archive for multiple targets.

```
$ MIX_ENV=prod mix elixir_make.precompile
Current compiling target: x86_64-apple-darwin
Fetching support libraries at https://example.com/support-darwin-x64.tar.gz
Compiles successfully...
Creating precompiled archive: /***-apple-darwin-0.15.1.tar.gz

Current compiling target: aarch64-apple-darwin
** Fails **
```

It fails because `make` won't be able to recognize the change in target, so it does not attempt to fetch a different package. Another issue I think is intermediate files (`*.o`) might be outdated, which we won't be able to catch here.

I think we should run `make clean` before compiling the [next target](https://github.com/elixir-lang/elixir_make/blob/master/lib/mix/tasks/elixir_make.precompile.ex#L35). That way, I can ensure to clean up by deleting all intermediate files.

Please let me know if you have suggestions to handle this differently.

**Comment by josevalim** [comment]
Any suggestions @cocoa-xu? I think we discussed using different directories per target?

**Comment by cocoa-xu** [comment]
Yes, we discussed that before, and we didn't go that way because common assets will be copied/downloaded multiple times.

> make clean before compiling the [next target](https://github.com/elixir-lang/elixir_make/blob/master/lib/mix/tasks/elixir_make.precompile.ex#L35).

I agree that we should call `make clean` before the next one if the Makefile contains a `clean` target (we can check the return code of `make --dry-run clean`, and it will be `2` if there is no such target). 

If there is no such target, should we 

- a) delete the `priv` dir and start fresh, or 
- b) keep it as is, try to compile the next target, and inform the developer about this `make clean` thing?

**Comment by josevalim** [comment]
I think checking --dry-run may be too brittle? Like, I am not sure if that works for nmake or gmake. Perhaps cc_precompiler could have a configuration to invoke a command before precompiling. If that option/config is set, you call it. It can even be set to `make`. WDYT?

**Comment by cocoa-xu** [comment]
Yep, that sounds better. 

But we still need another optional callback, say, `post_target_archive(target)` in `elixir_make` to notify the precompiler to do any cleanup work. Because currently the precompiler will only do the compilation work, and we can't cleanup before the current target is properly archived by `elixir_make`.

Ideally, I'm thinking of the following workflow in `elixir_make.precompile` (and possible optional callbacks)

1. `all_supported_targets(:compile)`. Required. 
2. `precompile(args, target)`. Required.
  
    It's already there. The precompiler will compile the project for a specific target.
    
3. `create_precompiled_archive/3`. 
    
    It's already there, a private function in `Mix.Tasks.ElixirMake.Precompile`.
    
    This archives the artefacts for the compiled target.
    
4. `post_target_archive(target)`. Optional.
    
    In this optional callback, the precompiler can do the cleanup job, like deleting all target-specific files.

**Comment by josevalim** [comment]
Why do we need to make 3 a callback? Or could we keep it as is?

**Comment by cocoa-xu** [comment]
Oh, I wrote it down because I want to make this workflow clearer, and it also explains why we need that `post_target_archive` callback. :D

**Comment by cocoa-xu** [comment]
Hi @akash-akya, `elixir_make` and `cc_precompiler` is updated, and it's now possible to cleanup job before the next target.

Let's say that there is a `clean` target in the makefile, then you need to set the `cleanup` key in `cc_precompiler`'s configuration. 

`make clean` will be called each time a precompilation target is done.

```elixir
def project do
  [
    # ...
    cc_precompiler: [
      cleanup: "clean"
    ],
    # ...
  ]
end
```

It will find and use the `make` executable the same way as `elixir_make` does: basically I copied the code from `elixir_make`, [cleanup/2](https://github.com/cocoa-xu/cc_precompiler/blob/682a2b2c0a8e7813db802cde05c03f6cf89888c7/lib/cc_precompiler.ex#L387-L522). 

So `make_executable`, `make_makefile`, `make_env` and variables like `MIX_TARGET` and `MIX_APP_PATH` will be effective and visible when cc_precompiler calls `make clean`.

**Comment by akash-akya** [comment]
Thanks @cocoa-xu for speedy fix! This solves my issue.

I configured an alias to avoid potential issue for the first platform during precompile, due to artifacts of previous operation.

```elixir
  def project do
    [
      cc_precompiler: [
        cleanup: "deep_clean"
      ],
      aliases: aliases()
      ...
    ]
  end

  defp aliases do
    [
      precompile: [
        "cmd make deep_clean",
        "elixir_make.precompile" 
      ]
    ]
  end
```
